Part of Issue2099: Only create Coinbase tx every payoutPeriod

hewison-chris opened the pull request:
Extracted this change to a separate PR as it is quite a big change and needs some tests to be fixed.
The `tx fees` for the validators and the Commons Budget will be accumulated until the payout block. 
In the payout block a single `Coinbase tx` has outputs to all non slashed validators and the Commons Budget.

hewison-chris committed:
Coinbase transactions only at each `payoutPeriod`

A `Coinbase` transaction is only created every payout block as defined
by `ConsensusParams.PayoutPeriod` number of blocks. Previously only
Validator payouts were using this interval, which meant that when adding
fees to transactions we would create Coinbase transactions every block
which paid to the Commons Budget. Now the fees for both validators and
the Commons Budget are accumulated until the payout block.

omerfirmak commented:
I thought fees already worked that way

https://github.com/bosagora/agora/blob/a838c60212f66b3b75fdc3322b3f42952edcc60d/source/agora/consensus/Fee.d#L411

hewison-chris commented:
> I thought fees already worked that way
> 
> https://github.com/bosagora/agora/blob/a838c60212f66b3b75fdc3322b3f42952edcc60d/source/agora/consensus/Fee.d#L411

When I updated TxBuilder to include fees based on the byte size of the `tx` I was seeing `Coinbase` transactions every block.

omerfirmak commented:
> 
> 
> > I thought fees already worked that way
> > https://github.com/bosagora/agora/blob/a838c60212f66b3b75fdc3322b3f42952edcc60d/source/agora/consensus/Fee.d#L411
> 
> When I updated TxBuilder to include fees based on the byte size of the `tx` I was seeing `Coinbase` transactions every block.

Yeah, now that I look at it; we pay commons budget every block.

omerfirmak commented:
We would need to somehow accumulate the fees we pay to the commons budget too, right? I think we would be burning coins this way.

hewison-chris commented:
> We would need to somehow accumulate the fees we pay to the commons budget too, right? I think we would be burning coins this way.

Yes I will add a test to check that the fees paid to the `Validators` and the `Commons Budget` are the correct `Amount` at the payout block. If we are not accumulating correctly this will catch it.

hewison-chris committed:
Trivial remove obsolete comment

hewison-chris committed:
Be more informative of reason for tx rejection

kchulj commented:
Tested with faucet, and these commits are compatible.

hewison-chris commented:
The test to check that no coins are lost due to paying out at the `PayoutPeriod` has been added.

hewison-chris commented:
I will add a temporary default fee in TxBuilder to prevent zero value fees causing problems with payout block `Coinbase tx`.

kchulj commented:
> I will add a temporary default fee in TxBuilder to prevent zero value fees causing problems with payout block `Coinbase tx`.

Is this represented by `Amount(0)`?

hewison-chris commented:
> > I will add a temporary default fee in TxBuilder to prevent zero value fees causing problems with payout block `Coinbase tx`.
> 
> Is this represented by `Amount(0)`?

The problem is we do not allow zero output transactions (which is good) but until we add fee calculation the Coinbase tx will have no outputs as we only add non zero (i.e. value > Amount(0)) outputs. I just realized though that temporarily allowing Coinbase tx with Amount(0) outputs will be easier as a work around until the TxBuilder updates in PR https://github.com/bosagora/agora/pull/2370 which will calculate the fees.

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -556,43 +556,34 @@ public class Ledger
         if (auto payouts = this.fee_man.getAccumulatedFees(height))
             foreach (pair; payouts.byKeyValue())
-                if (pair.value > Amount(0))
+                // TODO: uncomment when fees are calculated in TxBuilder
+                // if (pair.value > Amount(0))
```
Temporary until fees are calculated in TxBuilder

hewison-chris replied:
Is removed in https://github.com/bosagora/agora/pull/2370 which has to be merged after this one.

hewison-chris commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -86,7 +86,8 @@ public string isInvalidReason (
 
         // disallow 0 amount
-        if (output.value == Amount(0))
+        // TODO: remove Coinbase check when fees are calculated in TxBuilder
+        if (output.type != OutputType.Coinbase && output.value == Amount(0))
```
Temporary until fees are calculated in TxBuilder

hewison-chris replied:
Is updated in https://github.com/bosagora/agora/pull/2370 which has to be merged after this one.

hewison-chris committed:
Remove obsolete unit test

The test was checking that a block was not created when there are no txs
in the transaction pool. Now we can have empty blocks the test does not
make sense any more.

omerfirmak commented:
Overall, LGTM. Should I take a second look after it is all green?

hewison-chris commented:
All green

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -334,8 +334,10 @@ public class Ledger
-                reason !is null ? reason : "double-spend/coinbase", tx);
+            log.info("Rejected tx. Reason: {}. Tx: {}, txHash: {}",
+                reason !is null ? reason
+                    : tx.isCoinbase ? "Coinbase transaction" : "double-spend",
```
If we have nested ternaries, I'd say they warrant usage of parenthesis. But we can trivially avoid nested ternary via:
```D
if (!reason.length)
    reason = tx.isCoinbase ? "coinbase transaction" : "double spend";
```

hewison-chris replied:
I'll update in other PR as separate commit.

linked0 reviewed (approved):
LGTM

Geod24 reviewed (approved):
LGTM, just one style note but not worth blocking merge over it.
